File: packages/midi/android/src/main/java/expo/modules/midi/MidiModule.kt
package expo.modules.midi

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class MidiModule : Module() {
    private val midiManager by lazy { MidiManager(appContext.reactContext!!) }
    private var isSetup = false

    private fun ensureSetup() {
        if (isSetup) return
        midiManager.setup()
        isSetup = true

        midiManager.onMessage = { data ->
            sendEvent("onMidiMessage", mapOf(
                "data" to data.map { it.toInt() and 0xFF },
                "deltaTime" to 0
            ))
        }
        midiManager.onDevicesChanged = {
            sendEvent("onDevicesChanged", emptyMap<String, Any>())
        }
        midiManager.onDisconnect = {
            sendEvent("onDisconnect", emptyMap<String, Any>())
        }
    }

    override fun definition() = ModuleDefinition {
        Name("Midi")

        Events("onMidiMessage", "onDevicesChanged", "onDisconnect")

        OnStartObserving {
            ensureSetup()
        }

        OnDestroy {
            midiManager.onMessage = null
            midiManager.onDevicesChanged = null
            midiManager.onDisconnect = null
            midiManager.teardown()
        }

        Function("listInputs") {
            ensureSetup()
            midiManager.listInputs()
        }

        Function("listOutputs") {
            ensureSetup()
            midiManager.listOutputs()
        }

        AsyncFunction("openPorts") { inputName: String, outputName: String, promise: Promise ->
            ensureSetup()
            midiManager.openPorts(inputName, outputName) { success, error ->
                if (success) {
                    promise.resolve(null)
                } else {
                    promise.reject("MIDI_ERROR", error ?: "Failed to open ports", null)
                }
            }
        }

        Function("sendMessage") { data: List<Int> ->
            midiManager.send(data.map { it.toByte() }.toByteArray())
        }

        Function("closePorts") {
            midiManager.close()
        }
    }
}

File: packages/midi/ios/MidiModule.swift
import ExpoModulesCore

public class MidiModule: Module {
    private lazy var midiManager = MidiManager()
    private var isSetup = false

    private func ensureSetup() {
        guard !isSetup else { return }
        do {
            try midiManager.setup()
            isSetup = true

            midiManager.onMessage = { [weak self] data in
                let hex = data.prefix(10).map { String(format: "%02x", $0) }.joined(separator: " ")
                print("[MidiModule] sendEvent onMidiMessage: \(data.count)B, \(hex)")
                self?.sendEvent("onMidiMessage", [
                    "data": data.map { Int($0) },
                    "deltaTime": 0
                ])
            }
            midiManager.onDevicesChanged = { [weak self] in
                self?.sendEvent("onDevicesChanged", [:])
            }
            midiManager.onDisconnect = { [weak self] in
                self?.sendEvent("onDisconnect", [:])
            }
        } catch {
            print("MIDI setup failed: \(error)")
        }
    }

    public func definition() -> ModuleDefinition {
        Name("Midi")

        Events("onMidiMessage", "onDevicesChanged", "onDisconnect")

        // OnStartObserving intentionally empty — ensureSetup() is called by
        // each Function handler. This avoids a race condition where
        // OnStartObserving runs async on a background queue while openPorts
        // runs ensureSetup() synchronously, potentially double-creating MIDI ports.
        OnStartObserving {}

        OnDestroy {
            self.midiManager.onMessage = nil
            self.midiManager.onDevicesChanged = nil
            self.midiManager.onDisconnect = nil
            self.midiManager.disconnect()
        }

        Function("listInputs") { () -> [String] in
            self.ensureSetup()
            return self.midiManager.listSources()
        }

        Function("listOutputs") { () -> [String] in
            self.ensureSetup()
            return self.midiManager.listDestinations()
        }

        Function("openPorts") { (inputName: String, outputName: String) in
            self.ensureSetup()
            try self.midiManager.connect(sourceName: inputName, destName: outputName)
        }

        Function("sendMessage") { (data: [Int]) in
            let bytes = data.map { UInt8(clamping: $0) }
            self.midiManager.send(data: bytes)
        }

        Function("closePorts") {
            self.midiManager.disconnect()
        }
    }
}
